cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -397,6 +397,136 @@ struct PowerChecked {
+    static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status*) { \
+      static_assert(std::is_same<T, Arg0>::value, "");                                 \
+      return func(val);                                                                \
+    }                                                                                  \
+  };                                                                                   \
+  struct Trig##name##Checked {                                                         \
+    template <typename T, typename Arg0>                                               \
+    static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {        \
+      static_assert(std::is_same<T, double>::value, "");                               \
+      return func(val);                                                                \
+    }                                                                                  \
+    template <typename T, typename Arg0>                                               \
+    static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val,            \
+                                                  Status* st) {                        \
+      static_assert(std::is_same<T, Arg0>::value, "");                                 \
+      if (ARROW_PREDICT_FALSE(std::isinf(val))) {                                      \
+        *st = Status::Invalid("domain error");                                         \
+        return val;                                                                    \
+      }                                                                                \
+      return func(val);                                                                \
+    }                                                                                  \
+  };
+
+TRIG_NO_INF(Sin, std::sin);
+TRIG_NO_INF(Cos, std::cos);
+TRIG_NO_INF(Tan, std::tan);
+
+struct TrigAsin {
+  template <typename T, typename Arg0>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    return std::asin(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    return std::asin(val);
+  }
+};
+
+struct TrigAsinChecked {
+  template <typename T, typename Arg0>
+  static enable_if_integer<Arg0, T> Call(KernelContext*, Arg0 val, Status*) {
+    static_assert(std::is_same<T, double>::value, "");
+    return std::asin(val);
+  }
+
+  template <typename T, typename Arg0>
+  static enable_if_floating_point<Arg0, T> Call(KernelContext*, Arg0 val, Status* st) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE((val < -1.0 || val > 1.0) && !std::isnan(val))) {
```

Comment:
I suggest reorder the domain error checks as follows:
```
if (!std::isnan(val) && (val > 1.0 || val < -1.0)) { ... }
```
The reasoning is that a NaN input will exit immediately and a non-NaN input will be correctly compared to `|val| > 1` (C++ does not guarantees NaN comparisons to be false). Reordering the positive check before the negative check can make compiler reorder instructions more optimal, see [examples of all the combinations for clang and GCC](https://godbolt.org/z/PraGvMWr3).